{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build *Bacillus subtilis*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will build an ME model from the NC_000913.2 Genbank file, the iJO1366\n",
    "M model, and the complex reconstruction from iJL1678-ME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/Downloads/cobrapy-0.5.11/cobra/io/sbml3.py:24: UserWarning: Install lxml for faster SBML I/O\n",
      "  warn(\"Install lxml for faster SBML I/O\")\n",
      "/home/jt/Downloads/cobrapy-0.5.11/cobra/io/__init__.py:12: UserWarning: cobra.io.sbml requires libsbml\n",
      "  warn(\"cobra.io.sbml requires libsbml\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/__init__.py\n",
      "/home/jt/UCSD/bacillusme-master/ecolime/__init__.py\n"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function, division, absolute_import\n",
    "\n",
    "import sys\n",
    "sys.path.append('/home/jt/UCSD/bacillusme-master/')\n",
    "\n",
    "# python imports\n",
    "import re\n",
    "from os.path import join\n",
    "from collections import defaultdict\n",
    "import pickle\n",
    "\n",
    "# third party imports\n",
    "import pandas\n",
    "import cobra\n",
    "\n",
    "# COBRAme\n",
    "import cobrame\n",
    "from cobrame.util import building, mu, me_model_interface\n",
    "from cobrame.io.json import save_json_me_model, save_reduced_json_me_model\n",
    "\n",
    "# ECOLIme\n",
    "import ecolime\n",
    "from ecolime import (transcription, translation, flat_files, generics, formulas, compartments)\n",
    "from ecolime.util.helper_functions import *\n",
    "\n",
    "print(cobrame.__file__)\n",
    "print(ecolime.__file__)\n",
    "\n",
    "def print_rxn_info(me):\n",
    "    last = 0\n",
    "    reactionNumber = len(me.reactions);\n",
    "    for i in range(last,reactionNumber):\n",
    "        rxn = me.reactions[i]\n",
    "        print(i)\n",
    "        print(rxn)\n",
    "        try:\n",
    "            print(rxn.reaction)\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define Models\n",
    "eco_directory = join(flat_files.ecoli_files_dir, 'iJO1366.json')\n",
    "ijo_directory = join(flat_files.ecoli_files_dir, 'iYO844.json')\n",
    "uni_directory = join(flat_files.ecoli_files_dir, 'universal_model.json')\n",
    "\n",
    "eco = cobra.io.load_json_model(eco_directory)\n",
    "ijo = cobra.io.load_json_model(ijo_directory)\n",
    "uni = cobra.io.load_json_model(uni_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## M-model Fixes\n",
    "ijo.reactions.BIOMASS_BS_10.lower_bound = 0\n",
    "ijo.reactions.BIOMASS_BS_10.upper_bound = 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Create minimum solveable ME-model\n",
    "This will include the bare minimum representations of \n",
    "- Transcription Reactions\n",
    "- Translation Reactions\n",
    "- Complex Formation Reactions\n",
    "- Metabolic Reactions\n",
    "\n",
    "that still produce a working ME-model\n",
    "\n",
    "### 1) Create Model Object and populate its global info\n",
    "This includes important parameters that are used to calculate coupling \n",
    "constraints as well as organism-specific information such as peptide \n",
    "processing types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "me = cobrame.MEModel('iYO844-ME')\n",
    "\n",
    "# ME-models require special OptLang interface if cobrapy version >= 0.6.0\n",
    "# If cannot import SymbolicParameter, assume using cobrapy\n",
    "# versions <= 0.5.11\n",
    "try:\n",
    "    from optlang.interface import SymbolicParameter\n",
    "except:\n",
    "    pass\n",
    "else:\n",
    "    me.solver = me_model_interface\n",
    "\n",
    "# \"Translational capacity\" of organism\n",
    "me.global_info['kt'] = 4.5  # (in h-1)scott 2010, RNA-to-protein curve fit\n",
    "me.global_info['r0'] = 0.087  # scott 2010, RNA-to-protein curve fit\n",
    "me.global_info['k_deg'] = 1.0/5. * 60.0  # 1/5 1/min 60 min/h # h-1\n",
    "\n",
    "# Molecular mass of RNA component of ribosome\n",
    "me.global_info['m_rr'] = 1453. # in kDa\n",
    "\n",
    "# Average molecular mass of an amino acid\n",
    "me.global_info['m_aa'] = 109. / 1000.  # in kDa\n",
    "\n",
    "# Proportion of RNA that is rRNA\n",
    "me.global_info['f_rRNA'] = .86\n",
    "me.global_info['m_nt'] = 324. / 1000.  # in kDa\n",
    "me.global_info['f_mRNA'] = .02\n",
    "\n",
    "# tRNA associated global information\n",
    "me.global_info['m_tRNA'] = 25000. / 1000.  # in kDA\n",
    "me.global_info['f_tRNA'] = .12\n",
    "\n",
    "# Folding Properties\n",
    "me.global_info['temperature'] = 37\n",
    "me.global_info['propensity_scaling'] = .45\n",
    "\n",
    "# DNA Replication Parameters\n",
    "me.global_info['GC_fraction'] = 0.507896997096\n",
    "\n",
    "# Define the types of biomass that will be synthesized in the model\n",
    "me.add_biomass_constraints_to_model([\"protein_biomass\", \"mRNA_biomass\", \"tRNA_biomass\", \"rRNA_biomass\",\n",
    "                                     \"ncRNA_biomass\", \"DNA_biomass\", \"lipid_biomass\", \"constituent_biomass\",\n",
    "                                     \"prosthetic_group_biomass\", \"peptidoglycan_biomass\"])\n",
    "\n",
    "# Define ME-model compartments\n",
    "me.compartments = {\"p\": \"Periplasm\", \"e\": \"Extra-organism\", \"c\": \"Cytosol\", \n",
    "                   \"im\": 'Inner Membrane', 'om': \"Outer Membrane\", \"mc\": \"ME-model Constraint\",\n",
    "                   \"m\": \"Membrane\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2) Load metabolites and build Metabolic reactions\n",
    "The below reads in:\n",
    "- Required \n",
    " * **reaction_matrix.txt** (reaction matrix w/ reactions unlumped, metabolites renamed etc.)\n",
    " * **metabolites.txt** (metabolite properties)\n",
    " * **reactions.txt** (info on reversiblity, whether enzyme catalyzed etc.)\n",
    " * **m_to_me_mets.csv** (mapping of enzymes/complexes used in M-model to their ME-model compatible ID)\n",
    " * **protein_complexes.txt** (protein subunit stoichiometry of all complexes, used to identify metabolites as such)\n",
    " \n",
    "It creates a new e coli M-model from this info then incorporates it into the ME-model using *add_m_model_content*. metabolites are added directly reactions are added as StoichiometricData\n",
    "\n",
    "Metabolite types have different properties in an ME-model so enzyme complexes need added to the model as Complexes not Metabolites. Components in the E. coli M-model that are actually Complexes are compiled in *complex_list*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/bacillusme-master/ecolime/flat_files.py:60 \u001b[1;31mUserWarning\u001b[0m: Complex (BSU34550-MONOMER) in complex_stoichiometry_file twice\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BSU15920-MONOMER_mod_pan4p\n",
      "BSU15920-MONOMER\n",
      "BSU28500-MONOMER_mod_Oxidized\n",
      "BSU28500-MONOMER\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/bacillusme-master/ecolime/corrections.py:28 \u001b[1;31mUserWarning\u001b[0m: Creating new metabolite (4fe4s_c)\n",
      "/home/jt/UCSD/bacillusme-master/ecolime/corrections.py:28 \u001b[1;31mUserWarning\u001b[0m: Creating new metabolite (2fe2s_c)\n"
     ]
    }
   ],
   "source": [
    "# m_model = flat_files.get_m_model()\n",
    "m_model = flat_files.process_m_model(ijo, 'metabolites.txt',\n",
    "                                     'm_to_me_mets.csv', 'reactions.txt',\n",
    "                                     'reaction_matrix.txt',\n",
    "                                     'protein_complexes.txt',\n",
    "                                     defer_to_rxn_matrix=set())\n",
    "m_model.reactions.EX_glc__D_e.id = 'EX_glc__D_e'\n",
    "m_model.repair()\n",
    "# some of the \"metabolites\" in iJO1366 \"M\" model are actually complexes. We pass those in\n",
    "# so they get created as complexes, not metabolites.\n",
    "complexes = flat_files.get_complex_subunit_stoichiometry(\n",
    "    'protein_complexes.txt').keys()\n",
    "complex_set = set(\n",
    "    [i.id for i in m_model.metabolites if i.id.split('_mod_')[0] in complexes])\n",
    "building.add_m_model_content(me, m_model, complex_metabolite_ids=complex_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unknown metabolite '3fe4s' created\n",
      "unknown metabolite 'cl_c' created\n",
      "unknown metabolite 'cu_c' created\n",
      "unknown metabolite 'bmocogdp_c' created\n",
      "unknown metabolite 'dpm_c' created\n",
      "unknown metabolite 'cbl1_c' created\n"
     ]
    }
   ],
   "source": [
    "# This adds exchange reactions for metabolites not contained in iJO1366\n",
    "# Some of these cannot be produced by the model so they are added here\n",
    "exchange_list = ['4fe4s_c','2fe2s_c','udcpp_c','mn2_c','pydx5p_c','3fe4s','cl_c','sheme_c','cu_c',\n",
    "                 'bmocogdp_c','dpm_c','thmpp_c','zn2_c','cbl1_c','cobalt2_c']\n",
    "\n",
    "for met_id in exchange_list:\n",
    "    r = cobrame.MEReaction(\"EX_\" + met_id)\n",
    "    me.add_reaction(r)\n",
    "    r.reaction = met_id + \" <=> \""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3) Add Transcription and Translation \n",
    "The below reads in:\n",
    "- Required\n",
    " * **NC_000913.2.gb** (Genbank sequence annotation)\n",
    " * **ecolime/translation.py** (codon to tRNA mapping)\n",
    "- Optional\n",
    " * **TUs_from_ecocyc.txt** (TU definitions, start/stop positions, etc.)\n",
    " * **ecolime/translation.py** (dictionary of gene to frameshift mutation)\n",
    " \n",
    "To construct the bare minimimum components of a transcription and translation reactions. For example, transcription reactions at this point include nucleotides and the synthesized RNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "element_types = {'CDS', 'rRNA','tRNA', 'ncRNA','misc_RNA'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BSU_misc_RNA_1 as ncRNA\n",
      "Processing BSU_misc_RNA_2 as ncRNA\n",
      "Processing BSU_misc_RNA_3 as ncRNA\n",
      "Processing BSU_misc_RNA_4 as ncRNA\n",
      "Processing BSU_misc_RNA_5 as ncRNA\n",
      "Processing BSU_misc_RNA_6 as ncRNA\n",
      "Processing BSU_misc_RNA_64 as ncRNA\n",
      "Processing BSU_misc_RNA_7 as ncRNA\n",
      "Processing BSU_misc_RNA_65 as ncRNA\n",
      "Processing BSU_misc_RNA_66 as ncRNA\n",
      "Processing BSU_misc_RNA_8 as ncRNA\n",
      "Processing BSU_misc_RNA_9 as ncRNA\n",
      "Processing BSU_misc_RNA_10 as ncRNA\n",
      "Processing BSU_misc_RNA_11 as ncRNA\n",
      "Processing BSU_misc_RNA_12 as ncRNA\n",
      "Processing BSU_misc_RNA_13 as ncRNA\n",
      "Processing BSU_misc_RNA_67 as ncRNA\n",
      "Processing BSU_misc_RNA_14 as ncRNA\n",
      "Processing BSU_misc_RNA_15 as ncRNA\n",
      "Processing BSU_misc_RNA_16 as ncRNA\n",
      "Processing BSU_misc_RNA_17 as ncRNA\n",
      "Processing BSU_misc_RNA_18 as ncRNA\n",
      "Processing BSU_misc_RNA_19 as ncRNA\n",
      "Processing BSU_misc_RNA_68 as ncRNA\n",
      "Processing BSU_misc_RNA_87 as ncRNA\n",
      "Processing BSU_misc_RNA_20 as ncRNA\n",
      "Processing BSU_misc_RNA_21 as ncRNA\n",
      "Processing BSU_misc_RNA_22 as ncRNA\n",
      "Processing BSU_misc_RNA_88 as ncRNA\n",
      "Processing BSU_misc_RNA_69 as ncRNA\n",
      "Processing BSU_misc_RNA_70 as ncRNA\n",
      "Processing BSU_misc_RNA_23 as ncRNA\n",
      "Processing BSU_misc_RNA_24 as ncRNA\n",
      "Processing BSU_misc_RNA_25 as ncRNA\n",
      "Processing BSU_misc_RNA_26 as ncRNA\n",
      "Processing BSU_misc_RNA_27 as ncRNA\n",
      "Processing BSU_misc_RNA_28 as ncRNA\n",
      "Processing BSU_misc_RNA_29 as ncRNA\n",
      "Processing BSU_misc_RNA_30 as ncRNA\n",
      "Processing BSU_misc_RNA_71 as ncRNA\n",
      "Processing BSU_misc_RNA_89 as ncRNA\n",
      "Processing BSU_misc_RNA_72 as ncRNA\n",
      "Processing BSU_misc_RNA_73 as ncRNA\n",
      "Processing BSU_misc_RNA_31 as ncRNA\n",
      "Processing BSU_misc_RNA_74 as ncRNA\n",
      "Processing BSU_misc_RNA_75 as ncRNA\n",
      "Processing BSU_misc_RNA_32 as ncRNA\n",
      "Processing BSU_misc_RNA_76 as ncRNA\n",
      "Processing BSU_misc_RNA_33 as ncRNA\n",
      "Processing BSU_misc_RNA_90 as ncRNA\n",
      "Processing BSU_misc_RNA_77 as ncRNA\n",
      "Processing BSU_misc_RNA_34 as ncRNA\n",
      "Processing BSU_misc_RNA_35 as ncRNA\n",
      "Processing BSU_misc_RNA_78 as ncRNA\n",
      "Processing BSU_misc_RNA_36 as ncRNA\n",
      "Processing BSU_misc_RNA_91 as ncRNA\n",
      "Processing BSU_misc_RNA_37 as ncRNA\n",
      "Processing BSU_misc_RNA_39 as ncRNA\n",
      "Processing BSU_misc_RNA_79 as ncRNA\n",
      "Processing BSU_misc_RNA_80 as ncRNA\n",
      "Processing BSU_misc_RNA_81 as ncRNA\n",
      "Processing BSU_misc_RNA_92 as ncRNA\n",
      "Processing BSU_misc_RNA_82 as ncRNA\n",
      "Processing BSU_misc_RNA_40 as ncRNA\n",
      "Processing BSU_misc_RNA_41 as ncRNA\n",
      "Processing BSU_misc_RNA_42 as ncRNA\n",
      "Processing BSU_misc_RNA_43 as ncRNA\n",
      "Processing BSU_misc_RNA_44 as ncRNA\n",
      "Processing BSU_misc_RNA_45 as ncRNA\n",
      "Processing BSU_misc_RNA_46 as ncRNA\n",
      "Processing BSU_misc_RNA_47 as ncRNA\n",
      "Processing BSU_misc_RNA_48 as ncRNA\n",
      "Processing BSU_misc_RNA_93 as ncRNA\n",
      "Processing BSU_misc_RNA_83 as ncRNA\n",
      "Processing BSU_misc_RNA_49 as ncRNA\n",
      "Processing BSU_misc_RNA_50 as ncRNA\n",
      "Processing BSU_misc_RNA_51 as ncRNA\n",
      "Processing BSU_misc_RNA_52 as ncRNA\n",
      "Processing BSU_misc_RNA_84 as ncRNA\n",
      "Processing BSU_misc_RNA_53 as ncRNA\n",
      "Processing BSU_misc_RNA_85 as ncRNA\n",
      "Processing BSU_misc_RNA_54 as ncRNA\n",
      "Processing BSU_misc_RNA_55 as ncRNA\n",
      "Processing BSU_misc_RNA_56 as ncRNA\n",
      "Processing BSU_misc_RNA_57 as ncRNA\n",
      "Processing BSU_misc_RNA_58 as ncRNA\n",
      "Processing BSU_misc_RNA_59 as ncRNA\n",
      "Processing BSU_misc_RNA_60 as ncRNA\n",
      "Processing BSU_misc_RNA_61 as ncRNA\n",
      "Processing BSU_misc_RNA_62 as ncRNA\n",
      "Processing BSU_misc_RNA_63 as ncRNA\n",
      "Processing BSU_misc_RNA_86 as ncRNA\n",
      "Processing BSU_misc_RNA_94 as ncRNA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/reaction.py:852 \u001b[1;31mUserWarning\u001b[0m: RNA Polymerase () not found\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/reaction.py:1093 \u001b[1;31mUserWarning\u001b[0m: ribosome not found\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/reaction.py:1136 \u001b[1;31mUserWarning\u001b[0m: RNA_degradosome not found\n"
     ]
    }
   ],
   "source": [
    "gb_filename = join(flat_files.ecoli_files_dir,'NC_000964.gb')                                                                                    \n",
    "tu_df = flat_files.get_tu_dataframe('TUs_from_bsubcyc.txt')\n",
    "\n",
    "building.build_reactions_from_genbank(me, gb_filename, tu_df, verbose=False,\n",
    "                                      frameshift_dict=translation.frameshift_dict,\n",
    "                                      trna_to_codon=translation.trna_to_codon,\n",
    "                                      element_types = element_types)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<TranslationReaction translation_BSU22660 at 0x7f2ba37ef9e8>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me.reactions.get_by_id('translation_BSU22660')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4) Add in complex Formation without modifications (for now)\n",
    "\n",
    "The below reads in:\n",
    "- Required\n",
    " * **protein_complexes.txt** (Metabolic complexes' protein subunit stoichiometries)\n",
    " * **protein_modification.txt** (Type and number of modifications for each protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created <TranslatedGene protein_BSU29679 at 0x7f2ba6fff4a8> in <ComplexFormation formation_BSU29679-MONOMER at 0x7f2ba6fff550>\n",
      "Created <TranslatedGene protein_G8J2-7 at 0x7f2ba5b916a0> in <ComplexFormation formation_MONOMER8J2-7 at 0x7f2ba5b91be0>\n",
      "Created <TranslatedGene protein_G8J2-20 at 0x7f2ba5b50438> in <ComplexFormation formation_MONOMER8J2-49 at 0x7f2ba5b503c8>\n",
      "Created <TranslatedGene protein_G8J2-176 at 0x7f2ba5bcde48> in <ComplexFormation formation_MONOMER8J2-60 at 0x7f2ba5bcdda0>\n",
      "Created <TranslatedGene protein_G8J2-19 at 0x7f2ba38d7588> in <ComplexFormation formation_MONOMER8J2-48 at 0x7f2ba38d73c8>\n",
      "Created <TranslatedGene protein_G8J2-18 at 0x7f2ba38d75f8> in <ComplexFormation formation_MONOMER8J2-47 at 0x7f2ba38d7898>\n",
      "Created <TranslatedGene protein_G8J2-16 at 0x7f2ba38d7a20> in <ComplexFormation formation_MONOMER8J2-46 at 0x7f2ba38d74e0>\n",
      "Created <TranslatedGene protein_G8J2-13 at 0x7f2ba38d7940> in <ComplexFormation formation_MONOMER8J2-45 at 0x7f2ba38d79e8>\n",
      "Created <TranslatedGene protein_G8J2-173 at 0x7f2ba387a518> in <ComplexFormation formation_MONOMER8J2-58 at 0x7f2ba387a780>\n",
      "Created <TranslatedGene protein_G8J2-175 at 0x7f2ba387a588> in <ComplexFormation formation_MONOMER8J2-59 at 0x7f2ba387a470>\n",
      "Created <TranslatedGene protein_G8J2-21 at 0x7f2ba387a898> in <ComplexFormation formation_MONOMER8J2-50 at 0x7f2ba387ad30>\n",
      "Created <TranslatedGene protein_G8J2-162 at 0x7f2ba387ab38> in <ComplexFormation formation_MONOMER8J2-52 at 0x7f2ba387ac88>\n",
      "Created <TranslatedGene protein_G8J2-163 at 0x7f2ba387e160> in <ComplexFormation formation_MONOMER8J2-55 at 0x7f2ba387e1d0>\n",
      "Created <TranslatedGene protein_G8J2-170 at 0x7f2ba387e048> in <ComplexFormation formation_MONOMER8J2-56 at 0x7f2ba387e0f0>\n",
      "Created <TranslatedGene protein_G8J2-171 at 0x7f2ba387e588> in <ComplexFormation formation_MONOMER8J2-57 at 0x7f2ba387e208>\n",
      "Created <TranslatedGene protein_BSU23060 at 0x7f2ba38864e0> in <ComplexFormation formation_BSU23060-MONOMER at 0x7f2ba3886438>\n",
      "Created <TranslatedGene protein_BSU07540 at 0x7f2ba3836c50> in <ComplexFormation formation_BSU07540-MONOMER at 0x7f2ba3836b70>\n"
     ]
    }
   ],
   "source": [
    "# complex_stoichiometry_dict is a dict of {'complex_id': [{'bnum' : count}]}\n",
    "rna_components = {'BSU_misc_RNA_2','BSU_misc_RNA_35'} # component id should have 'RNA_ instead' of 'protein_'\n",
    "# get Metabolic Complex composition from ECOLIme\n",
    "complex_stoichiometry_dict = flat_files.get_complex_subunit_stoichiometry('protein_complexes.txt', rna_components)\n",
    "# add complexes to model\n",
    "complex_modification_dict = flat_files.get_complex_modifications('protein_modification.txt', 'protein_complexes.txt')\n",
    "building.add_model_complexes(me, complex_stoichiometry_dict, complex_modification_dict, verbose=False)\n",
    "\n",
    "# remove modifications. they will be added back in later\n",
    "for data in me.complex_data:\n",
    "    data.subreactions = {}\n",
    "# add formation reactions for each of the ComplexDatas\n",
    "for cplx_data in me.complex_data:\n",
    "    formation = cplx_data.formation\n",
    "    if formation:\n",
    "        formation.update()\n",
    "    else:\n",
    "        cplx_data.create_complex_formation()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5) Add dummy reaction to model and unmodeled_protein_fraction\n",
    "\n",
    "Includes the transcription, translation, complex_formation, and metabolic reaction. Sequence based on prevelance of each codon found in *E. coli*.\n",
    " - Required\n",
    "     * [**codon_usage.csv**](http://openwetware.org/wiki/Escherichia_coli/Codon_usage) (codon prevelance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = \"ATG\"\n",
    "codons = pandas.read_csv(join(flat_files.ecoli_files_dir, \"codon_usage.csv\"), index_col=0)\n",
    "for codon, row in codons.iterrows():\n",
    "    if row.amino_acid == \"Stop\":\n",
    "        continue\n",
    "    seq += codon * int(row.per_1000 // 3)  # want roughly 300 aa\n",
    "# get the most used stop codon\n",
    "seq += codons[codons.amino_acid == \"Stop\"].sort_values(\"per_1000\").index[-1]\n",
    "building.add_dummy_reactions(me, seq, update=True)\n",
    "\n",
    "# dummy proteins can be used to catalyze orphan reactions. This\n",
    "# ensures that orphan reactions will not becore favored to fulfil \n",
    "# unmodeled protein fraction requirement \n",
    "rxn = cobrame.SummaryVariable('dummy_protein_to_mass')\n",
    "me.add_reactions([rxn])\n",
    "mass = me.metabolites.protein_dummy.formula_weight / 1000.  # in kDa\n",
    "rxn.add_metabolites({'protein_biomass': -mass, 'protein_dummy': -1,\n",
    "                     cobrame.Constraint('unmodeled_protein_biomass'): mass})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6) Assocated Complexes and build Metabolic Reactions \n",
    "- Required\n",
    "    * **enzyme_reaction_association.txt**\n",
    "    * **reactions.txt** (gives reaction name, reversiblity, source and whether reaction is spontaneous)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/util/building.py:717 \u001b[1;31mUserWarning\u001b[0m: Reaction (BIOMASS_BS_10) cannot carry flux\n"
     ]
    }
   ],
   "source": [
    "# associate reaction id with the old ME complex id (including modifications)\n",
    "rxn_to_cplx_dict = flat_files.get_reaction_to_complex(m_model)\n",
    "rxn_info = flat_files.get_reaction_info_frame('reactions.txt')\n",
    "\n",
    "# Required to add dummy reaction as spontaneous reaction\n",
    "rxn_info = rxn_info.append(pandas.Series({'description': 'dummy reaction', 'is_reversible': 0,\n",
    "                                          'is_spontaneous':1}, name='dummy_reaction'))\n",
    "\n",
    "building.add_reactions_from_stoichiometric_data(me, rxn_to_cplx_dict, rxn_info, update=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7) Incorporate remaining biomass constituents\n",
    "There are leftover components from the *i*JO1366 biomass equation that either: \n",
    "1. have no mechanistic function in the model (glycogen)\n",
    "2. are cofactors that are regenerated (nad)\n",
    "\n",
    "Applies demands and coefficients from the *i*JO1366 biomass objective function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/model.py:103 \u001b[1;31mUserWarning\u001b[0m: Adding ATPM reaction to model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/model.py:86 \u001b[1;31mUserWarning\u001b[0m: Adding GAM reaction to model\n"
     ]
    }
   ],
   "source": [
    "me.ngam = 9.\n",
    "me.gam = 105.053483\n",
    "me.unmodeled_protein_fraction = .36\n",
    "\n",
    "biomass_constituents = {\n",
    "    \"10fthf_c\": -0.000367,\n",
    "    \"peptido_BS_c\": -.101817,\n",
    "    \"psetha_BS_c\": -0.00056,\n",
    "    \"mql7_c\":-0.000266,\n",
    "}\n",
    "\n",
    "\n",
    "rxn = cobrame.SummaryVariable('biomass_constituent_demand')\n",
    "me.add_reactions([rxn])\n",
    "rxn.add_metabolites(biomass_constituents)\n",
    "constituent_mass = sum(me.metabolites.get_by_id(c).formula_weight / 1000. * -v\n",
    "                     for c, v in biomass_constituents.items())\n",
    "rxn.lower_bound = mu\n",
    "rxn.upper_bound = mu\n",
    "rxn.add_metabolites({me.metabolites.constituent_biomass: constituent_mass})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " #### Lipid components\n",
    " Metabolites and coefficients from *i*JO1366 biomass objective function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify lipid metabolites in biomass equation\n",
    "\n",
    "lipid_demand = {}\n",
    "for key, value in ijo.reactions.BIOMASS_BS_10.metabolites.items():\n",
    "    if '_BS_c' in key.id and not key.id in biomass_constituents:\n",
    "        lipid_demand[key.id] = abs(value)\n",
    "                \n",
    "for met, requirement in lipid_demand.items():\n",
    "    component_mass = me.metabolites.get_by_id(met).formula_weight / 1000.\n",
    "    rxn = cobrame.SummaryVariable('Demand_' + met)\n",
    "    me.add_reactions([rxn])\n",
    "    rxn.add_metabolites({met: -1 * requirement,\n",
    "                         'lipid_biomass': component_mass * requirement})\n",
    "    rxn.lower_bound = mu\n",
    "    rxn.upper_bound = 1000.\n",
    "\n",
    "# Kdo2lipid4\n",
    "#requirement = 0.01945  # in mmol/gDW\n",
    "#met = me.metabolites.get_by_id('kdo2lipid4_e')\n",
    "#component_mass = met.formula_weight / 1000.\n",
    "#rxn = cobrame.SummaryVariable('Demand_' + met.id)\n",
    "#me.add_reactions([rxn])\n",
    "\n",
    "#rxn.add_metabolites({met.id : -1.*requirement,\n",
    "#                     'lipid_biomass': component_mass * requirement})\n",
    "#rxn.lower_bound = mu\n",
    "#rxn.upper_bound = mu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DNA Demand Requirements\n",
    "Added based on growth rate dependent DNA levels as in [O'brien EJ et al 2013](https://www.ncbi.nlm.nih.gov/pubmed/24084808)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_demand_stoich, dna_demand_bound = ecolime.dna_replication.return_gr_dependent_dna_demand(\n",
    "    me, me.global_info['GC_fraction'])\n",
    "\n",
    "dna_replication = cobrame.SummaryVariable(\"DNA_replication\")\n",
    "me.add_reaction(dna_replication)\n",
    "dna_replication.add_metabolites(dna_demand_stoich)\n",
    "dna_biomass = cobrame.Constraint(\"DNA_biomass\")\n",
    "\n",
    "dna_mw = 0\n",
    "dna_mw_no_ppi = ecolime.dna_replication.get_dna_mw_no_ppi_dict(me)\n",
    "for met, value in me.reactions.DNA_replication.metabolites.items():\n",
    "    if met.id != 'ppi_c':\n",
    "        dna_mw -= value * dna_mw_no_ppi[met.id.replace('_c','')] / 1000.\n",
    "        \n",
    "dna_replication.add_metabolites({dna_biomass: dna_mw})\n",
    "dna_replication.lower_bound = dna_demand_bound \n",
    "dna_replication.upper_bound = dna_demand_bound \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**: From this point forward, executing every codeblock should result in a solveable ME-model\n",
    "\n",
    "------\n",
    "\n",
    "## Part 2: Add metastructures to solving ME-model\n",
    "This includes:\n",
    "1. ribosome\n",
    "2. RNA polymerase\n",
    "3. charged_tRNAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sometimes multiple entities can perform the same role. To prevent a combinatorial explosion of possibilities, we can create  \"generic\" version, where any of those entities can fill in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for generic, components in generics.generic_dict.items():\n",
    "    cobrame.GenericData(generic, me, components).create_reactions()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1) Add ribosome\n",
    "This uses the ribosome composition definition in **ecolime/ribosome.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (mg2_c) does not have formula\n"
     ]
    }
   ],
   "source": [
    "ecolime.ribosome.add_ribosome(me, verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2) Add charged tRNA reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tRNA charging reactions were automatically added when loading the genome from the genbank file. However, the charging reactions still need to be made aware of the tRNA synthetases which are responsible.\n",
    "\n",
    "Uses **trna_charging.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "aa_synthetase_dict = ecolime.trna_charging.amino_acid_trna_synthetase\n",
    "for data in me.tRNA_data:\n",
    "    data.synthetase = str(aa_synthetase_dict[data.amino_acid])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generic charged tRNAs are added to translation reactions via SubreactionData below.\n",
    "\n",
    "All new data added in this block contained in **ecolime/translation.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:870 \u001b[1;31mUserWarning\u001b[0m: initiation subreaction Translation_initiation_factor_InfA not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:870 \u001b[1;31mUserWarning\u001b[0m: initiation subreaction Translation_initiation_factor_InfC not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:870 \u001b[1;31mUserWarning\u001b[0m: initiation subreaction fmet_addition_at_START not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:870 \u001b[1;31mUserWarning\u001b[0m: initiation subreaction Translation_gtp_initiation_factor_InfB not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:836 \u001b[1;31mUserWarning\u001b[0m: elongation subreaction Tuf_gtp_regeneration not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:836 \u001b[1;31mUserWarning\u001b[0m: elongation subreaction FusA_mono_elongation not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU05343 starts with 'UGC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU05343\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU05344 starts with 'GCU' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU05890 starts with 'UUU' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU05899\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU06033 starts with 'GAG' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU06033\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU06074\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU06076 starts with 'UUA' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU06076\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU06078 starts with 'UUU' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU06079 starts with 'GAC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU06079\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU06083\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU07170\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU11381\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU11382 starts with 'GGC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU12680 starts with 'AUC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU16890\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU16900 starts with 'AGG' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU17266\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU19810\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:801 \u001b[1;31mUserWarning\u001b[0m: tRNA addition subreaction sec_addition_at_UGA not in model\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU20040\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU20060 starts with 'CAA' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU20060\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU21599 starts with 'AAU' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU21670 starts with 'UCU' which is not a start codon\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding selenocystein for BSU20040\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU23330\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU23836 starts with 'UGC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU23836\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU26039 starts with 'AGC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU26040\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU26748 starts with 'AUA' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU26749\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU26750\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU32960 starts with 'GGC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU32970\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU35290 starts with 'GAC' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU35290\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU35609 starts with 'UGG' which is not a start codon\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:900 \u001b[1;31mUserWarning\u001b[0m: No termination enzyme for RNA_BSU35610\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:863 \u001b[1;31mUserWarning\u001b[0m: RNA_BSU39029 starts with 'AAA' which is not a start codon\n"
     ]
    }
   ],
   "source": [
    "ecolime.translation.add_charged_trna_subreactions(me)\n",
    "for data in me.translation_data:\n",
    "    data.add_initiation_subreactions(start_codons=translation.translation_start_codons,\n",
    "                                    start_subreactions=set(translation.initiation_subreactions.keys()))\n",
    "\n",
    "    data.add_elongation_subreactions(elongation_subreactions=set(translation.elongation_subreactions.keys()))\n",
    "\n",
    "    data.add_termination_subreactions(translation_terminator_dict=translation.translation_stop_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3) Add Transcription Metacomplexes\n",
    "#### RNA Polymerase\n",
    "\n",
    "Data for RNA_polymerase composition fround in **ecolime/transcription**\n",
    "\n",
    "Uses *tu_df* from **TUs_from_ecocyc.txt**, above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in transcription.rna_polymerases:\n",
    "    rnap_obj = cobrame.RNAP(met)\n",
    "    me.add_metabolites(rnap_obj)\n",
    "transcription.add_rna_polymerase_complexes(me, verbose=False)\n",
    "\n",
    "# associate the correct RNA_polymerase and factors to TUs\n",
    "sigma_to_rnap_dict = transcription.sigma_factor_complex_to_rna_polymerase_dict\n",
    "for tu_id in tu_df.index:\n",
    "    transcription_data = me.process_data.get_by_id(tu_id)\n",
    "    sigma = tu_df.sigma[tu_id]\n",
    "    rna_polymerase = sigma_to_rnap_dict[sigma]\n",
    "    transcription_data.RNA_polymerase = rna_polymerase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Degradosome (both for RNA degradation and RNA splicing)\n",
    "\n",
    "All new data contained in **transcription.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "me.add_metabolites([cobrame.Complex('RNA_degradosome')])\n",
    "data = cobrame.ComplexData('RNA_degradosome', me)\n",
    "for subunit, value in transcription.rna_degradosome.items():\n",
    "    data.stoichiometry[subunit] = value\n",
    "data.create_complex_formation(verbose=False)\n",
    "\n",
    "# Used for RNA splicing\n",
    "data = cobrame.SubreactionData('RNA_degradation_machine', me)\n",
    "data.enzyme = 'RNA_degradosome'\n",
    "\n",
    "data = cobrame.SubreactionData('RNA_degradation_atp_requirement', me)\n",
    "# .25 water equivaltent for atp hydrolysis per nucleotide\n",
    "data.stoichiometry = {'atp_c': -.25, 'h2o_c': -.25, 'adp_c': .25,\n",
    "                      'pi_c': .25, 'h_c': .25}\n",
    "\n",
    "transcription.add_rna_splicing(me)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------\n",
    "## Part 3: Add remaining modifications\n",
    "rRNA modifications handled in *add_ribosome*\n",
    "\n",
    "### 1) Add complex modifications\n",
    "*complex_modification_dict* from **protein_modification.text**, above\n",
    "\n",
    "The rest of the new data contained in **ecolime/modifications.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "for complex_id, info in complex_modification_dict.items():\n",
    "    modifications = {}\n",
    "    for mod, value in info['modifications'].items():\n",
    "        # stoichiometry of modification determined in \n",
    "        # subreaction_data.stoichiometry\n",
    "        \n",
    "        modifications['mod_' + mod] = abs(value)\n",
    "    me.process_data.get_by_id(complex_id).subreactions = modifications\n",
    "\n",
    "# Adds modification data for more complicated enzyme modifications\n",
    "# (ie, iron sulfur cluster modification)\n",
    "ecolime.modifications.add_modification_procedures(me)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./me_models/iBS_checkpoint.pickle\", \"wb\") as outfile:\n",
    "    pickle.dump(me, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./me_models/iBS_checkpoint.pickle\", \"rb\") as outfile:\n",
    "    me = pickle.load(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created <Metabolite 3fe4s_c at 0x7f2ba280c128> in <ComplexFormation formation_CPLX8J2-2_mod_4fe4s_mod_3fe4s_mod_bmocogdp at 0x7f2b978f0128>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (cu_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (mn2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (zn2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (cbl1_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (ca2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (cobalt2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (ni2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (fe2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (pheme_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (dpm_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (cu2_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (cl_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (sheme_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (3fe4s_c) does not have formula\n",
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (bmocogdp_c) does not have formula\n"
     ]
    }
   ],
   "source": [
    "# add formation reactions for each of the ComplexDatas\n",
    "for cplx_data in me.complex_data:\n",
    "    formation = cplx_data.formation\n",
    "    if 'Oxidized' not in cplx_data.id: \n",
    "        if formation:\n",
    "            formation.update()\n",
    "        else:\n",
    "            cplx_data.create_complex_formation()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2) Add tRNA mods and asocciate them with tRNA charging reactions\n",
    "New data from:\n",
    "1. **ecolime/trna_charging.py** (read via *add_tRNA_modification_procedures()*)\n",
    "2. **post_transcriptional_modification_of_tRNA.txt** (modification types per tRNA)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add tRNA modifications to ME-model\n",
    "ecolime.trna_charging.add_trna_modification_procedures(me)\n",
    "\n",
    "# trna_modifications = {tRNA_id: {modifications: count}}\n",
    "trna_modifications=flat_files.get_trna_modification_targets()\n",
    "for trna in trna_modifications:\n",
    "    for data in me.process_data.query(trna):\n",
    "        data.subreactions = trna_modifications[trna]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Part 4: Add remaining subreactions\n",
    "### 1) Add translation related subreactions\n",
    "All new data from **ecolime/translation.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding selenocystein for BSU20040\n"
     ]
    }
   ],
   "source": [
    "# add the translation subreaction data objects to model\n",
    "translation.add_translation_subreactions_to_model(me)\n",
    "\n",
    "# add translation subreaction data to reactions\n",
    "methionine_cleaved=translation.methionine_cleaved\n",
    "folding_dict=translation.folding_dict\n",
    "\n",
    "for data in me.translation_data:\n",
    "    locus_id = data.id\n",
    "    if locus_id in methionine_cleaved:\n",
    "        data.subreactions['N_terminal_methionine_cleavage'] = 1\n",
    "\n",
    "    for folding_type in folding_dict:\n",
    "        if locus_id in folding_dict[folding_type]:\n",
    "            data.subreactions[folding_type] = 1\n",
    "\n",
    "    # This block was ran above, but should be ran again to \n",
    "    # incorporate any subreactions not added previously\n",
    "    data.add_initiation_subreactions(\n",
    "        start_codons=translation.translation_start_codons,\n",
    "        start_subreactions=set(translation.initiation_subreactions.keys()))\n",
    "    data.add_elongation_subreactions(\n",
    "        elongation_subreactions=set(translation.elongation_subreactions.keys()))\n",
    "    data.add_termination_subreactions(\n",
    "        translation_terminator_dict=translation.translation_stop_dict)\n",
    "\n",
    "    # add organism specific subreactions associated with peptide processing\n",
    "    for subrxn in translation.peptide_processing_subreactions:\n",
    "        data.subreactions[subrxn] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2) Add transcription related subreactions\n",
    "All new data from **ecolime/transcription.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subreaction in transcription.transcription_subreactions:\n",
    "    subreaction_data = cobrame.SubreactionData(subreaction, me)\n",
    "    enzymes = transcription.transcription_subreactions[subreaction]['enzymes']\n",
    "    subreaction_data.stoichiometry = transcription.transcription_subreactions[subreaction]['stoich']\n",
    "    subreaction_data.enzyme = enzymes\n",
    "\n",
    "for transcription_data in me.transcription_data:\n",
    "    # Assume false if not in tu_df\n",
    "    rho_dependent = tu_df.rho_dependent.get(transcription_data.id, False)\n",
    "    rho = 'dependent' if rho_dependent else 'independent'\n",
    "    stable = 'stable' if transcription_data.codes_stable_rna else 'normal'\n",
    "    transcription_data.subreactions['Transcription_%s_rho_%s' % (stable,\n",
    "                                                                 rho)] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "## Part 5: Add in translocation\n",
    "\n",
    "New data from:\n",
    "1. **peptide_compartment_and_pathways.txt** (Protein compartment and translocation pathway for each membrane complex)\n",
    "2. **ecolime/translocation.py** (definitions of each translocation pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add TranslocationData\n",
    "transloc = pandas.read_csv(join(flat_files.ecoli_files_dir,\"peptide_compartment_and_pathways.txt\"),\n",
    "                           sep='\\t', comment=\"#\")\n",
    "for pathway, info in ecolime.translocation.pathway.items():\n",
    "    if 'alt' not in pathway:\n",
    "        transloc_data = cobrame.TranslocationData(pathway + '_translocation', me)\n",
    "    else:\n",
    "        transloc_data = cobrame.TranslocationData(pathway.replace('_alt', '_translocation_alt'), me)\n",
    "    transloc_data.enzyme_dict = info['enzymes']\n",
    "    transloc_data.keff = info['keff']\n",
    "    transloc_data.length_dependent_energy = info['length_dependent_energy']\n",
    "    transloc_data.stoichiometry = info['stoichiometry']\n",
    "\n",
    "# Associate data and add translocation reactions\n",
    "ecolime.translocation.add_translocation_pathways(me, transloc, membrane_constraints = False)\n",
    "\n",
    "# Update stoichiometry of membrane complexes\n",
    "# new_stoich = {complex_id: protein_w_compartment}\n",
    "new_stoich = defaultdict(dict)\n",
    "for cplx, row in transloc.set_index('Complex').iterrows():\n",
    "    if cplx == 'EG10544-MONOMER':\n",
    "        continue\n",
    "    protein = row.Protein.split('(')[0] + '_' + row.Protein_compartment\n",
    "    value = me.process_data.get_by_id(cplx).stoichiometry['protein_' + row.Protein.split('(')[0]]\n",
    "    new_stoich[cplx]['protein_' + protein] = float(value)\n",
    "\n",
    "for cplx, stoich in new_stoich.items():\n",
    "    complex_data = me.process_data.get_by_id(cplx)\n",
    "    for met, value in stoich.items():\n",
    "        complex_data.stoichiometry.pop(met[0:16])\n",
    "        complex_data.stoichiometry[met] = value\n",
    "        complex_data.formation.update()\n",
    "    # Complex ids in protein compartment file doesn't include mods\n",
    "    # Some have multiple alternative modifications so must loop through these\n",
    "    for complex_data in me.process_data.query(cplx + '_mod_'):\n",
    "        for met, value in stoich.items():\n",
    "            complex_data.stoichiometry.pop(met[0:16])\n",
    "            complex_data.stoichiometry[met] = value\n",
    "            complex_data.formation.update()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Part 6: Add Cell Wall Components\n",
    "All new data from **ecolime/translocation.py**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "compartment_dict = {}\n",
    "for prot, compartment in transloc.set_index('Protein').Protein_compartment.to_dict().items():\n",
    "    compartment_dict[prot.split('(')[0]] = compartment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add lipid modification SubreactionData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bsub m-model doest not have pe and pg metabolites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Braun's lipoprotein demand\n",
    "Metabolites and coefficients as defined in [Liu et al 2014](http://bmcsystbiol.biomedcentral.com/articles/10.1186/s12918-014-0110-6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correct complex formation IDs if they contain lipoproteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bsub m-model doest not have murein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "## Part 7: Set keffs\n",
    "\n",
    "Either entirely based on SASA or using fit keffs from [Ebrahim et al 2016](https://www.ncbi.nlm.nih.gov/pubmed/27782110?dopt=Abstract)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# # Set keffs to sasa fluxes centered around 65.\n",
    "# me.set_SASA_keffs(65)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BSU15920-MONOMER_mod_pan4p  has no formula\n",
      "BSU28500-MONOMER_mod_Oxidized  has no formula\n",
      "BSU29679-MONOMER  has no formula\n",
      "MONOMER8J2-7  has no formula\n",
      "MONOMER8J2-49  has no formula\n",
      "MONOMER8J2-60  has no formula\n",
      "MONOMER8J2-48  has no formula\n",
      "MONOMER8J2-47  has no formula\n",
      "MONOMER8J2-46  has no formula\n",
      "MONOMER8J2-45  has no formula\n",
      "MONOMER8J2-58  has no formula\n",
      "MONOMER8J2-59  has no formula\n",
      "MONOMER8J2-50  has no formula\n",
      "MONOMER8J2-52  has no formula\n",
      "MONOMER8J2-55  has no formula\n",
      "MONOMER8J2-56  has no formula\n",
      "MONOMER8J2-57  has no formula\n",
      "BSU23060-MONOMER  has no formula\n",
      "BSU07540-MONOMER  has no formula\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/core/Metabolite.py:97 \u001b[1;31mUserWarning\u001b[0m: The element 'X' does not appear in the peridic table\n"
     ]
    }
   ],
   "source": [
    "## Get median sasa\n",
    "from cobrame.core.component import Complex\n",
    "import numpy as np\n",
    "\n",
    "sasa_list = []\n",
    "for met in me.metabolites:\n",
    "    cplx_sasa = 0.\n",
    "    if not isinstance(met, Complex):\n",
    "        continue\n",
    "    MW = met.formula_weight\n",
    "    if not MW:\n",
    "        MW = 0\n",
    "        print(met.id, ' has no formula')\n",
    "    cplx_sasa += MW ** (3. / 4)\n",
    "    sasa_list.append(cplx_sasa)\n",
    "median_sasa = np.median(np.array(sasa_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "reaction_median_keffs = pandas.read_csv(flat_files.fixpath('reaction_median_keffs.txt'), sep='\\t',\n",
    "                              names=['Reaction', 'Complex', 'Median_Keff'])\n",
    "keff_list = []\n",
    "for key, row in reaction_median_keffs.iterrows():\n",
    "    reaction_id = row['Reaction']\n",
    "    complex_id = row['Complex']\n",
    "    median_keff = row['Median_Keff']\n",
    "    \n",
    "    real_reaction_id = reaction_id + ''\n",
    "    \n",
    "    molecular_weight = me.metabolites.get_by_id(complex_id).formula_weight\n",
    "    sasa = molecular_weight ** (3./4)\n",
    "    keff = sasa * median_keff / median_sasa\n",
    "    \n",
    "    if keff > 3000:\n",
    "        keff = 3000.\n",
    "    elif keff < .01:\n",
    "        keff = .01\n",
    "    keff_list.append(keff)\n",
    "    try:\n",
    "        me.reactions.get_by_id(reaction_id).keff = keff\n",
    "        me.reactions.get_by_id(reaction_id).update()\n",
    "    except:\n",
    "        continue\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "## Part 8: Model updates and corrections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add reaction subsystems from iJO to model\n",
    "for rxn in ijo.reactions:\n",
    "    if rxn.id in me.process_data:\n",
    "        data = me.process_data.get_by_id(rxn.id)\n",
    "    else:\n",
    "        continue\n",
    "    for r in data.parent_reactions:\n",
    "        r.subsystem = rxn.subsystem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add enzymatic coupling for \"carriers\"\n",
    "These are enzyme complexes that act as metabolites in a metabolic reaction (i.e. are metabolites in iJO1366)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "for data in me.stoichiometric_data:\n",
    "    if data.id == 'dummy_reaction':\n",
    "        continue\n",
    "        \n",
    "    for met, value in data.stoichiometry.items():\n",
    "        if not isinstance(me.metabolites.get_by_id(met), cobrame.Complex) or value > 0:\n",
    "            continue\n",
    "            \n",
    "        subreaction_id = met + '_carrier_activity'\n",
    "        if subreaction_id not in me.process_data:           \n",
    "            sub = cobrame.SubreactionData(met + '_carrier_activity', me)\n",
    "            sub.enzyme = met\n",
    "        data.subreactions[subreaction_id] = abs(value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "## Part 9: Update and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "me.reactions.biomass_dilution.objective_coefficient = 1.\n",
    "me.reactions.EX_glc__D_e.lower_bound = -1000\n",
    "me.reactions.EX_o2_e.lower_bound = -1000\n",
    "me.ngam = 9.\n",
    "me.gam = 105.053483\n",
    "me.unmodeled_protein_fraction = .36\n",
    "\n",
    "me.reactions.BIOMASS_BS_10_FWD_CPLX_dummy.lower_bound = 0\n",
    "me.reactions.BIOMASS_BS_10_FWD_CPLX_dummy.upper_bound = 0\n",
    "me.reactions.ATPM_FWD_CPLX_dummy.lower_bound = 0\n",
    "me.reactions.ATPM_FWD_CPLX_dummy.upper_bound = 0\n",
    "\n",
    "me.metabolites.get_by_id('mn2_c').formula = 'Mn2'\n",
    "me.metabolites.get_by_id('mg2_c').formula = 'Mg2'\n",
    "me.metabolites.get_by_id('ca2_c').formula = 'Ca2'\n",
    "me.metabolites.get_by_id('cu2_c').formula = 'Cu2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created <Metabolite Oxidized_c at 0x7f2b969f0cc0> in <ComplexFormation formation_BSU33480-MONOMER_mod_Oxidized at 0x7f2b97956ba8>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jt/UCSD/cobrame-master/cobrame/core/processdata.py:256 \u001b[1;31mUserWarning\u001b[0m: met (Oxidized_c) does not have formula\n"
     ]
    }
   ],
   "source": [
    "me.update()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "me.prune()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add remaining complex formulas and compartments to model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update a second time to incorporate all of the metabolite formulas corectly\n",
    "for r in me.reactions.query('formation_'):\n",
    "    r.update()\n",
    "\n",
    "# Update complex formulas\n",
    "df = pandas.read_table(join(flat_files.ecoli_files_dir, 'modification.txt'),\n",
    "                            names=['mod', 'formula','na'])\n",
    "df = df.drop('na', axis=1).set_index('mod').dropna(how='any')\n",
    "modification_formulas = df.T.to_dict()\n",
    "formulas.add_remaining_complex_formulas(me, modification_formulas)\n",
    "\n",
    "# Update reactions affected by formula update \n",
    "for r in me.reactions.query('_mod_lipo'):\n",
    "    r.update()\n",
    "for r in me.reactions.query('_mod_glycyl'):\n",
    "    r.update()\n",
    "    \n",
    "# add metabolite compartments\n",
    "compartments.add_compartments_to_model(me)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remaining fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BSU00090\n",
      "BSU00110\n",
      "BSU00120\n",
      "BSU00130\n",
      "BSU00140\n",
      "BSU00150\n",
      "BSU00180\n",
      "BSU00270\n",
      "BSU00280\n",
      "BSU00380\n",
      "BSU00410\n",
      "BSU00460\n",
      "BSU00500\n",
      "BSU00510\n",
      "BSU00550\n",
      "BSU00680\n",
      "BSU00730\n",
      "BSU00740\n",
      "BSU00750\n",
      "BSU00760\n",
      "BSU00770\n",
      "BSU00780\n",
      "BSU00790\n",
      "BSU00810\n",
      "BSU00820\n",
      "BSU00900\n",
      "BSU00910\n",
      "BSU00930\n",
      "BSU00940\n",
      "BSU00980\n",
      "BSU01000\n",
      "BSU01010\n",
      "BSU01070\n",
      "BSU01080\n",
      "BSU01100\n",
      "BSU01110\n",
      "BSU01120\n",
      "BSU01130\n",
      "BSU01150\n",
      "BSU01200\n",
      "BSU01220\n",
      "BSU01230\n",
      "BSU01250\n",
      "BSU01260\n",
      "BSU01290\n",
      "BSU01300\n",
      "BSU01310\n",
      "BSU01320\n",
      "BSU01330\n",
      "BSU01340\n",
      "BSU01350\n",
      "BSU01360\n",
      "BSU01370\n",
      "BSU01380\n",
      "BSU01390\n",
      "BSU01400\n",
      "BSU01410\n",
      "BSU01420\n",
      "BSU01430\n",
      "BSU01500\n",
      "BSU01530\n",
      "BSU01580\n",
      "BSU01610\n",
      "BSU01620\n",
      "BSU01630\n",
      "BSU01680\n",
      "BSU01730\n",
      "BSU01770\n",
      "BSU01780\n",
      "BSU02090\n",
      "BSU02130\n",
      "BSU02140\n",
      "BSU02230\n",
      "BSU02270\n",
      "BSU02290\n",
      "BSU02340\n",
      "BSU02350\n",
      "BSU02360\n",
      "BSU02390\n",
      "BSU02400\n",
      "BSU02420\n",
      "BSU02430\n",
      "BSU02460\n",
      "BSU02480\n",
      "BSU02490\n",
      "BSU02510\n",
      "BSU02630\n",
      "BSU02640\n",
      "BSU02750\n",
      "BSU02760\n",
      "BSU02850\n",
      "BSU02860\n",
      "BSU02870\n",
      "BSU02980\n",
      "BSU02990\n",
      "BSU03000\n",
      "BSU03010\n",
      "BSU03040\n",
      "BSU03050\n",
      "BSU03060\n",
      "BSU03070\n",
      "BSU03130\n",
      "BSU03150\n",
      "BSU03200\n",
      "BSU03220\n",
      "BSU03270\n",
      "BSU03280\n",
      "BSU03290\n",
      "BSU03300\n",
      "BSU03310\n",
      "BSU03320\n",
      "BSU03330\n",
      "BSU03790\n",
      "BSU03800\n",
      "BSU03810\n",
      "BSU03820\n",
      "BSU03830\n",
      "BSU03860\n",
      "BSU03900\n",
      "BSU03910\n",
      "BSU03930\n",
      "BSU03950\n",
      "BSU03981\n",
      "BSU03990\n",
      "BSU04000\n",
      "BSU04030\n",
      "BSU04340\n",
      "BSU04360\n",
      "BSU04470\n",
      "BSU04520\n",
      "BSU04560\n",
      "BSU04570\n",
      "BSU04620\n",
      "BSU04640\n",
      "BSU04730\n",
      "BSU04740\n",
      "BSU05340\n",
      "BSU05740\n",
      "BSU05810\n",
      "BSU05820\n",
      "BSU05830\n",
      "BSU05860\n",
      "BSU05900\n",
      "BSU05980\n",
      "BSU05990\n",
      "BSU06020\n",
      "BSU06030\n",
      "BSU06150\n",
      "BSU06160\n",
      "BSU06170\n",
      "BSU06230\n",
      "BSU06240\n",
      "BSU06310\n",
      "BSU06360\n",
      "BSU06370\n",
      "BSU06420\n",
      "BSU06430\n",
      "BSU06440\n",
      "BSU06450\n",
      "BSU06470\n",
      "BSU06480\n",
      "BSU06490\n",
      "BSU06500\n",
      "BSU06510\n",
      "BSU06520\n",
      "BSU06530\n",
      "BSU06660\n",
      "BSU06670\n",
      "BSU06680\n",
      "BSU06690\n",
      "BSU06730\n",
      "BSU06970\n",
      "BSU06980\n",
      "BSU06990\n",
      "BSU07490\n",
      "BSU07500\n",
      "BSU07510\n",
      "BSU07520\n",
      "BSU07570\n",
      "BSU07610\n",
      "BSU07700\n",
      "BSU07800\n",
      "BSU07810\n",
      "BSU07830\n",
      "BSU07840\n",
      "BSU07920\n",
      "BSU07990\n",
      "BSU08000\n",
      "BSU08030\n",
      "BSU08060\n",
      "BSU08070\n",
      "BSU08180\n",
      "BSU08200\n",
      "BSU08440\n",
      "BSU08450\n",
      "BSU08460\n",
      "BSU08650\n",
      "BSU08710\n",
      "BSU08790\n",
      "BSU08820\n",
      "BSU08830\n",
      "BSU08840\n",
      "BSU08850\n",
      "BSU08860\n",
      "BSU09130\n",
      "BSU09240\n",
      "BSU09280\n",
      "BSU09290\n",
      "BSU09300\n",
      "BSU09310\n",
      "BSU09410\n",
      "BSU09440\n",
      "BSU09520\n",
      "BSU09530\n",
      "BSU09540\n",
      "BSU09680\n",
      "BSU09840\n",
      "BSU09850\n",
      "BSU10010\n",
      "BSU10020\n",
      "BSU10120\n",
      "BSU10130\n",
      "BSU10140\n",
      "BSU10170\n",
      "BSU10220\n",
      "BSU10270\n",
      "BSU10360\n",
      "BSU10520\n",
      "BSU10790\n",
      "BSU10870\n",
      "BSU11010\n",
      "BSU11190\n",
      "BSU11200\n",
      "BSU11210\n",
      "BSU11220\n",
      "BSU11230\n",
      "BSU11240\n",
      "BSU11250\n",
      "BSU11330\n",
      "BSU11340\n",
      "BSU11420\n",
      "BSU11610\n",
      "BSU11640\n",
      "BSU11650\n",
      "BSU11670\n",
      "BSU11710\n",
      "BSU11720\n",
      "BSU11870\n",
      "BSU11880\n",
      "BSU12010\n",
      "BSU12020\n",
      "BSU12080\n",
      "BSU12160\n",
      "BSU12290\n",
      "BSU12300\n",
      "BSU12330\n",
      "BSU12340\n",
      "BSU12360\n",
      "BSU12380\n",
      "BSU12390\n",
      "BSU12840\n",
      "BSU12910\n",
      "BSU12920\n",
      "BSU12930\n",
      "BSU12940\n",
      "BSU12950\n",
      "BSU12960\n",
      "BSU13010\n",
      "BSU13110\n",
      "BSU13120\n",
      "BSU13130\n",
      "BSU13170\n",
      "BSU13180\n",
      "BSU13300\n",
      "BSU13450\n",
      "BSU13550\n",
      "BSU13560\n",
      "BSU13580\n",
      "BSU13590\n",
      "BSU13600\n",
      "BSU13610\n",
      "BSU13620\n",
      "BSU13730\n",
      "BSU13850\n",
      "BSU13890\n",
      "BSU13900\n",
      "BSU13910\n",
      "BSU13960\n",
      "BSU14000\n",
      "BSU14180\n",
      "BSU14190\n",
      "BSU14390\n",
      "BSU14400\n",
      "BSU14520\n",
      "BSU14530\n",
      "BSU14540\n",
      "BSU14580\n",
      "BSU14590\n",
      "BSU14600\n",
      "BSU14610\n",
      "BSU14630\n",
      "BSU14670\n",
      "BSU14730\n",
      "BSU14830\n",
      "BSU14860\n",
      "BSU14870\n",
      "BSU14880\n",
      "BSU14890\n",
      "BSU14900\n",
      "BSU14910\n",
      "BSU14920\n",
      "BSU15020\n",
      "BSU15080\n",
      "BSU15110\n",
      "BSU15140\n",
      "BSU15180\n",
      "BSU15190\n",
      "BSU15200\n",
      "BSU15220\n",
      "BSU15230\n",
      "BSU15320\n",
      "BSU15350\n",
      "BSU15430\n",
      "BSU15480\n",
      "BSU15490\n",
      "BSU15500\n",
      "BSU15510\n",
      "BSU15520\n",
      "BSU15530\n",
      "BSU15540\n",
      "BSU15550\n",
      "BSU15560\n",
      "BSU15570\n",
      "BSU15580\n",
      "BSU15590\n",
      "BSU15600\n",
      "BSU15610\n",
      "BSU15620\n",
      "BSU15630\n",
      "BSU15650\n",
      "BSU15680\n",
      "BSU15690\n",
      "BSU15700\n",
      "BSU15720\n",
      "BSU15730\n",
      "BSU15750\n",
      "BSU15790\n",
      "BSU15820\n",
      "BSU15850\n",
      "BSU15860\n",
      "BSU15900\n",
      "BSU15910\n",
      "BSU15920\n",
      "BSU15930\n",
      "BSU15980\n",
      "BSU15990\n",
      "BSU16020\n",
      "BSU16090\n",
      "BSU16100\n",
      "BSU16470\n",
      "BSU16490\n",
      "BSU16500\n",
      "BSU16510\n",
      "BSU16520\n",
      "BSU16530\n",
      "BSU16540\n",
      "BSU16550\n",
      "BSU16570\n",
      "BSU16600\n",
      "BSU16630\n",
      "BSU16650\n",
      "BSU16670\n",
      "BSU16680\n",
      "BSU16690\n",
      "BSU16730\n",
      "BSU16740\n",
      "BSU16750\n",
      "BSU16760\n",
      "BSU16770\n",
      "BSU16780\n",
      "BSU16920\n",
      "BSU16960\n",
      "BSU16990\n",
      "BSU17000\n",
      "BSU17150\n",
      "BSU17380\n",
      "BSU17390\n",
      "BSU17410\n",
      "BSU17460\n",
      "BSU17600\n",
      "BSU17610\n",
      "BSU17660\n",
      "BSU17680\n",
      "BSU17710\n",
      "BSU17890\n",
      "BSU18000\n",
      "BSU18120\n",
      "BSU18210\n",
      "BSU18220\n",
      "BSU18230\n",
      "BSU18250\n",
      "BSU18260\n",
      "BSU18410\n",
      "BSU18440\n",
      "BSU18450\n",
      "BSU18470\n",
      "BSU18480\n",
      "BSU18540\n",
      "BSU18570\n",
      "BSU18670\n",
      "BSU18800\n",
      "BSU18830\n",
      "BSU19230\n",
      "BSU19310\n",
      "BSU19350\n",
      "BSU19360\n",
      "BSU19370\n",
      "BSU19520\n",
      "BSU19580\n",
      "BSU19630\n",
      "BSU19690\n",
      "BSU19710\n",
      "BSU19800\n",
      "BSU20020\n",
      "BSU20250\n",
      "BSU21680\n",
      "BSU21690\n",
      "BSU21770\n",
      "BSU21810\n",
      "BSU21820\n",
      "BSU21870\n",
      "BSU21900\n",
      "BSU21910\n",
      "BSU21920\n",
      "BSU22060\n",
      "BSU22070\n",
      "BSU22090\n",
      "BSU22100\n",
      "BSU22110\n",
      "BSU22130\n",
      "BSU22140\n",
      "BSU22360\n",
      "BSU22370\n",
      "BSU22410\n",
      "BSU22420\n",
      "BSU22430\n",
      "BSU22440\n",
      "BSU22480\n",
      "BSU22490\n",
      "BSU22540\n",
      "BSU22550\n",
      "BSU22560\n",
      "BSU22600\n",
      "BSU22610\n",
      "BSU22620\n",
      "BSU22630\n",
      "BSU22640\n",
      "BSU22650\n",
      "BSU22660\n",
      "BSU22670\n",
      "BSU22680\n",
      "BSU22690\n",
      "BSU22700\n",
      "BSU22710\n",
      "BSU22730\n",
      "BSU22740\n",
      "BSU22750\n",
      "BSU22760\n",
      "BSU22780\n",
      "BSU22830\n",
      "BSU22870\n",
      "BSU22890\n",
      "BSU22960\n",
      "BSU23070\n",
      "BSU23080\n",
      "BSU23100\n",
      "BSU23250\n",
      "BSU23260\n",
      "BSU23270\n",
      "BSU23280\n",
      "BSU23380\n",
      "BSU23450\n",
      "BSU23490\n",
      "BSU23500\n",
      "BSU23560\n",
      "BSU23570\n",
      "BSU23580\n",
      "BSU23610\n",
      "BSU23760\n",
      "BSU23770\n",
      "BSU23800\n",
      "BSU23840\n",
      "BSU23850\n",
      "BSU23860\n",
      "BSU23920\n",
      "BSU23960\n",
      "BSU23970\n",
      "BSU23980\n",
      "BSU24030\n",
      "BSU24040\n",
      "BSU24050\n",
      "BSU24060\n",
      "BSU24070\n",
      "BSU24080\n",
      "BSU24090\n",
      "BSU24120\n",
      "BSU24130\n",
      "BSU24140\n",
      "BSU24150\n",
      "BSU24160\n",
      "BSU24170\n",
      "BSU24270\n",
      "BSU24280\n",
      "BSU24310\n",
      "BSU24340\n",
      "BSU24350\n",
      "BSU24740\n",
      "BSU24850\n",
      "BSU24950\n",
      "BSU24960\n",
      "BSU24970\n",
      "BSU24980\n",
      "BSU24990\n",
      "BSU25020\n",
      "BSU25070\n",
      "BSU25160\n",
      "BSU25200\n",
      "BSU25260\n",
      "BSU25270\n",
      "BSU25290\n",
      "BSU25300\n",
      "BSU25310\n",
      "BSU25410\n",
      "BSU25420\n",
      "BSU25500\n",
      "BSU25550\n",
      "BSU25580\n",
      "BSU25640\n",
      "BSU25660\n",
      "BSU25710\n",
      "BSU25760\n",
      "BSU25790\n",
      "BSU25900\n",
      "BSU26390\n",
      "BSU26590\n",
      "BSU26600\n",
      "BSU26640\n",
      "BSU26650\n",
      "BSU26690\n",
      "BSU26810\n",
      "BSU26860\n",
      "BSU26970\n",
      "BSU27010\n",
      "BSU27040\n",
      "BSU27050\n",
      "BSU27060\n",
      "BSU27070\n",
      "BSU27120\n",
      "BSU27200\n",
      "BSU27220\n",
      "BSU27250\n",
      "BSU27260\n",
      "BSU27270\n",
      "BSU27320\n",
      "BSU27330\n",
      "BSU27410\n",
      "BSU27430\n",
      "BSU27440\n",
      "BSU27450\n",
      "BSU27460\n",
      "BSU27550\n",
      "BSU27560\n",
      "BSU27600\n",
      "BSU27610\n",
      "BSU27650\n",
      "BSU27810\n",
      "BSU27850\n",
      "BSU27860\n",
      "BSU27870\n",
      "BSU27900\n",
      "BSU27910\n",
      "BSU27940\n",
      "BSU28080\n",
      "BSU28090\n",
      "BSU28120\n",
      "BSU28130\n",
      "BSU28140\n",
      "BSU28150\n",
      "BSU28230\n",
      "BSU28250\n",
      "BSU28260\n",
      "BSU28270\n",
      "BSU28280\n",
      "BSU28290\n",
      "BSU28300\n",
      "BSU28310\n",
      "BSU28370\n",
      "BSU28390\n",
      "BSU28430\n",
      "BSU28440\n",
      "BSU28450\n",
      "BSU28470\n",
      "BSU28500\n",
      "BSU28510\n",
      "BSU28540\n",
      "BSU28560\n",
      "BSU28630\n",
      "BSU28640\n",
      "BSU28680\n",
      "BSU28720\n",
      "BSU28730\n",
      "BSU28740\n",
      "BSU28750\n",
      "BSU28780\n",
      "BSU28790\n",
      "BSU28800\n",
      "BSU28860\n",
      "BSU28870\n",
      "BSU28950\n",
      "BSU29010\n",
      "BSU29020\n",
      "BSU29050\n",
      "BSU29060\n",
      "BSU29120\n",
      "BSU29130\n",
      "BSU29140\n",
      "BSU29180\n",
      "BSU29190\n",
      "BSU29200\n",
      "BSU29210\n",
      "BSU29220\n",
      "BSU29300\n",
      "BSU29340\n",
      "BSU29350\n",
      "BSU29360\n",
      "BSU29370\n",
      "BSU29380\n",
      "BSU29440\n",
      "BSU29450\n",
      "BSU29470\n",
      "BSU29600\n",
      "BSU29620\n",
      "BSU29660\n",
      "BSU29670\n",
      "BSU29680\n",
      "BSU29750\n",
      "BSU29790\n",
      "BSU29990\n",
      "BSU30070\n",
      "BSU30200\n",
      "BSU30210\n",
      "BSU30220\n",
      "BSU30230\n",
      "BSU30240\n",
      "BSU30300\n",
      "BSU30320\n",
      "BSU30410\n",
      "BSU30420\n",
      "BSU30430\n",
      "BSU30440\n",
      "BSU30450\n",
      "BSU30530\n",
      "BSU30540\n",
      "BSU30550\n",
      "BSU30560\n",
      "BSU30690\n",
      "BSU30740\n",
      "BSU30750\n",
      "BSU30760\n",
      "BSU30770\n",
      "BSU30780\n",
      "BSU30790\n",
      "BSU30800\n",
      "BSU30820\n",
      "BSU30830\n",
      "BSU30870\n",
      "BSU30940\n",
      "BSU30950\n",
      "BSU30960\n",
      "BSU30970\n",
      "BSU30980\n",
      "BSU31050\n",
      "BSU31060\n",
      "BSU31090\n",
      "BSU31100\n",
      "BSU31150\n",
      "BSU31180\n",
      "BSU31200\n",
      "BSU31322\n",
      "BSU31350\n",
      "BSU31400\n",
      "BSU31580\n",
      "BSU31600\n",
      "BSU31610\n",
      "BSU31620\n",
      "BSU31630\n",
      "BSU31640\n",
      "BSU31650\n",
      "BSU31660\n",
      "BSU31750\n",
      "BSU31760\n",
      "BSU31930\n",
      "BSU31970\n",
      "BSU31980\n",
      "BSU31990\n",
      "BSU32000\n",
      "BSU32100\n",
      "BSU32110\n",
      "BSU32130\n",
      "BSU32170\n",
      "BSU32180\n",
      "BSU32200\n",
      "BSU32240\n",
      "BSU32250\n",
      "BSU32260\n",
      "BSU32410\n",
      "BSU32430\n",
      "BSU32440\n",
      "BSU32450\n",
      "BSU32460\n",
      "BSU32470\n",
      "BSU32480\n",
      "BSU32490\n",
      "BSU32500\n",
      "BSU32510\n",
      "BSU32680\n",
      "BSU32820\n",
      "BSU32830\n",
      "BSU32840\n",
      "BSU33040\n",
      "BSU33160\n",
      "BSU33170\n",
      "BSU33180\n",
      "BSU33240\n",
      "BSU33290\n",
      "BSU33300\n",
      "BSU33310\n",
      "BSU33330\n",
      "BSU33370\n",
      "BSU33380\n",
      "BSU33390\n",
      "BSU33400\n",
      "BSU33420\n",
      "BSU33430\n",
      "BSU33440\n",
      "BSU33490\n",
      "BSU33500\n",
      "BSU33510\n",
      "BSU33630\n",
      "BSU33700\n",
      "BSU33710\n",
      "BSU33720\n",
      "BSU33730\n",
      "BSU33800\n",
      "BSU33810\n",
      "BSU33820\n",
      "BSU33830\n",
      "BSU33900\n",
      "BSU33910\n",
      "BSU33920\n",
      "BSU33930\n",
      "BSU33940\n",
      "BSU33960\n",
      "BSU34130\n",
      "BSU34190\n",
      "BSU34200\n",
      "BSU34550\n",
      "BSU34560\n",
      "BSU34570\n",
      "BSU34590\n",
      "BSU34600\n",
      "BSU34610\n",
      "BSU34660\n",
      "BSU34790\n",
      "BSU34860\n",
      "BSU34870\n",
      "BSU34880\n",
      "BSU34890\n",
      "BSU34900\n",
      "BSU34910\n",
      "BSU34920\n",
      "BSU35010\n",
      "BSU35020\n",
      "BSU35290\n",
      "BSU35300\n",
      "BSU35530\n",
      "BSU35540\n",
      "BSU35550\n",
      "BSU35570\n",
      "BSU35580\n",
      "BSU35590\n",
      "BSU35600\n",
      "BSU35609\n",
      "BSU35620\n",
      "BSU35660\n",
      "BSU35670\n",
      "BSU35680\n",
      "BSU35690\n",
      "BSU35700\n",
      "BSU35710\n",
      "BSU35720\n",
      "BSU35730\n",
      "BSU35740\n",
      "BSU35750\n",
      "BSU35760\n",
      "BSU35780\n",
      "BSU35790\n",
      "BSU35920\n",
      "BSU35930\n",
      "BSU35940\n",
      "BSU35950\n",
      "BSU35960\n",
      "BSU36000\n",
      "BSU36010\n",
      "BSU36030\n",
      "BSU36120\n",
      "BSU36130\n",
      "BSU36370\n",
      "BSU36470\n",
      "BSU36510\n",
      "BSU36590\n",
      "BSU36640\n",
      "BSU36650\n",
      "BSU36660\n",
      "BSU36710\n",
      "BSU36760\n",
      "BSU36800\n",
      "BSU36810\n",
      "BSU36820\n",
      "BSU36830\n",
      "BSU36840\n",
      "BSU36850\n",
      "BSU36860\n",
      "BSU36870\n",
      "BSU36880\n",
      "BSU36890\n",
      "BSU36900\n",
      "BSU36920\n",
      "BSU37010\n",
      "BSU37060\n",
      "BSU37070\n",
      "BSU37080\n",
      "BSU37100\n",
      "BSU37110\n",
      "BSU37120\n",
      "BSU37150\n",
      "BSU37160\n",
      "BSU37170\n",
      "BSU37190\n",
      "BSU37240\n",
      "BSU37320\n",
      "BSU37330\n",
      "BSU37490\n",
      "BSU37500\n",
      "BSU37660\n",
      "BSU37710\n",
      "BSU37760\n",
      "BSU37780\n",
      "BSU37790\n",
      "BSU38020\n",
      "BSU38040\n",
      "BSU38050\n",
      "BSU38060\n",
      "BSU38110\n",
      "BSU38140\n",
      "BSU38150\n",
      "BSU38160\n",
      "BSU38170\n",
      "BSU38190\n",
      "BSU38200\n",
      "BSU38240\n",
      "BSU38260\n",
      "BSU38270\n",
      "BSU38280\n",
      "BSU38290\n",
      "BSU38300\n",
      "BSU38390\n",
      "BSU38490\n",
      "BSU38500\n",
      "BSU38510\n",
      "BSU38520\n",
      "BSU38530\n",
      "BSU38550\n",
      "BSU38560\n",
      "BSU38570\n",
      "BSU38580\n",
      "BSU38590\n",
      "BSU38630\n",
      "BSU38700\n",
      "BSU38710\n",
      "BSU38750\n",
      "BSU38760\n",
      "BSU38770\n",
      "BSU38830\n",
      "BSU38860\n",
      "BSU38980\n",
      "BSU38990\n",
      "BSU39020\n",
      "BSU39050\n",
      "BSU39060\n",
      "BSU39260\n",
      "BSU39270\n",
      "BSU39350\n",
      "BSU39360\n",
      "BSU39370\n",
      "BSU39380\n",
      "BSU39390\n",
      "BSU39400\n",
      "BSU39410\n",
      "BSU39420\n",
      "BSU39450\n",
      "BSU39540\n",
      "BSU39610\n",
      "BSU39670\n",
      "BSU39700\n",
      "BSU39710\n",
      "BSU39720\n",
      "BSU39730\n",
      "BSU39740\n",
      "BSU39750\n",
      "BSU39760\n",
      "BSU39780\n",
      "BSU39860\n",
      "BSU39920\n",
      "BSU39980\n",
      "BSU40040\n",
      "BSU40060\n",
      "BSU40070\n",
      "BSU40110\n",
      "BSU40190\n",
      "BSU40320\n",
      "BSU40330\n",
      "BSU40340\n",
      "BSU40420\n",
      "BSU40550\n",
      "BSU40850\n",
      "BSU40890\n",
      "BSU40910\n",
      "BSU41050\n",
      "BSU01000\n",
      "BSU01360\n",
      "BSU01580\n",
      "BSU01610\n",
      "BSU01620\n",
      "BSU01630\n",
      "BSU02140\n",
      "BSU02270\n",
      "BSU02290\n",
      "BSU02340\n",
      "BSU02400\n",
      "BSU02420\n",
      "BSU02480\n",
      "BSU02630\n",
      "BSU02640\n",
      "BSU02980\n",
      "BSU02990\n",
      "BSU03000\n",
      "BSU03040\n",
      "BSU03060\n",
      "BSU03070\n",
      "BSU03220\n",
      "BSU03330\n",
      "BSU03950\n",
      "BSU04360\n",
      "BSU04470\n",
      "BSU05340\n",
      "BSU05980\n",
      "BSU05990\n",
      "BSU06160\n",
      "BSU06230\n",
      "BSU06310\n",
      "BSU06370\n",
      "BSU06660\n",
      "BSU07490\n",
      "BSU07500\n",
      "BSU07510\n",
      "BSU07520\n",
      "BSU07570\n",
      "BSU07610\n",
      "BSU07840\n",
      "BSU07920\n",
      "BSU08000\n",
      "BSU08830\n",
      "BSU08840\n",
      "BSU08850\n",
      "BSU09130\n",
      "BSU09280\n",
      "BSU09540\n",
      "BSU09680\n",
      "BSU09850\n",
      "BSU10010\n",
      "BSU10220\n",
      "BSU10520\n",
      "BSU10870\n",
      "BSU11640\n",
      "BSU12010\n",
      "BSU12080\n",
      "BSU12360\n",
      "BSU12840\n",
      "BSU12930\n",
      "BSU12940\n",
      "BSU12950\n",
      "BSU12960\n",
      "BSU13180\n",
      "BSU13300\n",
      "BSU13850\n",
      "BSU14400\n",
      "BSU14870\n",
      "BSU14880\n",
      "BSU14890\n",
      "BSU14900\n",
      "BSU14910\n",
      "BSU14920\n",
      "BSU15190\n",
      "BSU15220\n",
      "BSU15480\n",
      "BSU15580\n",
      "BSU15650\n",
      "BSU16540\n",
      "BSU16920\n",
      "BSU16960\n",
      "BSU17710\n",
      "BSU18120\n",
      "BSU18410\n",
      "BSU18540\n",
      "BSU18670\n",
      "BSU18800\n",
      "BSU19350\n",
      "BSU19520\n",
      "BSU19580\n",
      "BSU19800\n",
      "BSU21920\n",
      "BSU22060\n",
      "BSU22090\n",
      "BSU23560\n",
      "BSU24740\n",
      "BSU24950\n",
      "BSU24960\n",
      "BSU24970\n",
      "BSU24980\n",
      "BSU24990\n",
      "BSU25290\n",
      "BSU25310\n",
      "BSU25420\n",
      "BSU25790\n",
      "BSU26590\n",
      "BSU26640\n",
      "BSU26650\n",
      "BSU26690\n",
      "BSU26860\n",
      "BSU27200\n",
      "BSU27430\n",
      "BSU27440\n",
      "BSU27450\n",
      "BSU27460\n",
      "BSU27600\n",
      "BSU27650\n",
      "BSU27810\n",
      "BSU28430\n",
      "BSU28440\n",
      "BSU28450\n",
      "BSU28870\n",
      "BSU29340\n",
      "BSU29350\n",
      "BSU29360\n",
      "BSU29370\n",
      "BSU29380\n",
      "BSU29600\n",
      "BSU29990\n",
      "BSU30070\n",
      "BSU30530\n",
      "BSU30740\n",
      "BSU30750\n",
      "BSU30760\n",
      "BSU30770\n",
      "BSU31090\n",
      "BSU31100\n",
      "BSU31150\n",
      "BSU31322\n",
      "BSU31580\n",
      "BSU31600\n",
      "BSU31610\n",
      "BSU31620\n",
      "BSU31630\n",
      "BSU31640\n",
      "BSU31650\n",
      "BSU31660\n",
      "BSU32180\n",
      "BSU32430\n",
      "BSU32440\n",
      "BSU33240\n",
      "BSU33330\n",
      "BSU33420\n",
      "BSU33490\n",
      "BSU33500\n",
      "BSU33630\n",
      "BSU33700\n",
      "BSU33710\n",
      "BSU33720\n",
      "BSU33730\n",
      "BSU33800\n",
      "BSU33810\n",
      "BSU33820\n",
      "BSU33830\n",
      "BSU33900\n",
      "BSU33960\n",
      "BSU34190\n",
      "BSU34570\n",
      "BSU34590\n",
      "BSU34600\n",
      "BSU34610\n",
      "BSU34660\n",
      "BSU35300\n",
      "BSU35780\n",
      "BSU36030\n",
      "BSU36120\n",
      "BSU36130\n",
      "BSU36470\n",
      "BSU36510\n",
      "BSU36590\n",
      "BSU37190\n",
      "BSU37240\n",
      "BSU37320\n",
      "BSU37760\n",
      "BSU38060\n",
      "BSU38140\n",
      "BSU38150\n",
      "BSU38160\n",
      "BSU38170\n",
      "BSU38240\n",
      "BSU38490\n",
      "BSU38710\n",
      "BSU38750\n",
      "BSU38760\n",
      "BSU38770\n",
      "BSU39020\n",
      "BSU39060\n",
      "BSU39390\n",
      "BSU39410\n",
      "BSU39450\n",
      "BSU39710\n",
      "BSU40070\n",
      "BSU40330\n",
      "EX_4fe4s_c not in m_model\n",
      "EX_2fe2s_c not in m_model\n",
      "EX_udcpp_c not in m_model\n",
      "EX_mn2_c not in m_model\n",
      "EX_pydx5p_c not in m_model\n",
      "EX_3fe4s not in m_model\n",
      "EX_cl_c not in m_model\n",
      "EX_sheme_c not in m_model\n",
      "EX_cu_c not in m_model\n",
      "EX_bmocogdp_c not in m_model\n",
      "EX_dpm_c not in m_model\n",
      "EX_thmpp_c not in m_model\n",
      "EX_zn2_c not in m_model\n",
      "EX_cbl1_c not in m_model\n",
      "EX_cobalt2_c not in m_model\n"
     ]
    }
   ],
   "source": [
    "##\n",
    "current_gene_ids = [gene.id for gene in me.genes]\n",
    "for met in me.metabolites:\n",
    "    met_id = met.id\n",
    "    if 'protein_BSU' in met_id:\n",
    "        gene_id = met_id.split('_')[1]\n",
    "        if gene_id and gene_id not in current_gene_ids:\n",
    "            try:\n",
    "                print(gene_id)\n",
    "                gene = cobra.Gene(gene_id)\n",
    "                me.genes.append(gene)\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "##\n",
    "me.reactions.BIOMASS_BS_10_FWD_CPLX_dummy.lower_bound = 0\n",
    "me.reactions.BIOMASS_BS_10_FWD_CPLX_dummy.upper_bound = 0\n",
    "me.reactions.ATPM_FWD_CPLX_dummy.lower_bound = 0\n",
    "me.reactions.ATPM_FWD_CPLX_dummy.upper_bound = 0\n",
    "\n",
    "## Homogenize exchange reactions\n",
    "for rxn in me.reactions:\n",
    "    if 'EX_' in rxn.id:\n",
    "        try:\n",
    "            lb = ijo.reactions.get_by_id(rxn.id).lower_bound\n",
    "            ub = ijo.reactions.get_by_id(rxn.id).upper_bound\n",
    "            rxn.lower_bound = lb\n",
    "            rxn.upper_bound = ub\n",
    "        except:\n",
    "            print(rxn.id, 'not in m_model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./me_models/iYO844b.pickle\", \"wb\") as outfile:\n",
    "    pickle.dump(me, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./me_models/iYO844b.pickle\", \"rb\") as outfile:\n",
    "    me = pickle.load(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## New fixes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "( Demand_pgly_BS_c mu mu ) \t 0.000176 pgly_BS_c <=> 0.012548534412479996 lipid_biomass\n",
      "( Demand_lipo1_24_BS_c mu mu ) \t 7e-06 lipo1_24_BS_c <=> 0.00591416142296 lipid_biomass\n",
      "( Demand_d12dg_BS_c mu mu ) \t 0.00011 d12dg_BS_c <=> 0.0097263716968 lipid_biomass\n",
      "( Demand_gtca2_45_BS_c mu mu ) \t 0.002347 gtca2_45_BS_c <=> 0.026715503385342003 lipid_biomass\n",
      "( Demand_lysylpgly_BS_c mu mu ) \t 2.2e-05 lysylpgly_BS_c <=> 0.0018527632855599998 lipid_biomass\n",
      "( Demand_lipo3_24_BS_c mu mu ) \t 1.8e-05 lipo3_24_BS_c <=> 0.01205219511504 lipid_biomass\n",
      "( Demand_gtca1_45_BS_c mu mu ) \t 0.003624 gtca1_45_BS_c <=> 0.026722062249264 lipid_biomass\n",
      "( Demand_gtca3_45_BS_c mu mu ) \t 0.001819 gtca3_45_BS_c <=> 0.026717435194354 lipid_biomass\n",
      "( Demand_t12dg_BS_c mu mu ) \t 6.6e-05 t12dg_BS_c <=> 0.006905950978080001 lipid_biomass\n",
      "( Demand_m12dg_BS_c mu mu ) \t 8.6e-05 m12dg_BS_c <=> 0.00620984507568 lipid_biomass\n",
      "( Demand_cdlp_BS_c mu mu ) \t 5e-06 cdlp_BS_c <=> 0.0006669379998 lipid_biomass\n",
      "( Demand_tcam_BS_c mu mu ) \t 0.003112 tcam_BS_c <=> 0.04316222890296 lipid_biomass\n",
      "( Demand_lipo2_24_BS_c mu mu ) \t 6e-06 lipo2_24_BS_c <=> 0.005660428867679999 lipid_biomass\n",
      "( Demand_lipo4_24_BS_c mu mu ) \t 1.5e-05 lipo4_24_BS_c <=> 0.0068361414492 lipid_biomass\n"
     ]
    }
   ],
   "source": [
    "# Feb 12, 2020\n",
    "# Model was synthesizing extra lipids to compensate and reach maximum growth rate.\n",
    "# Fix: Set lb = ub = mass fraction of lipids\n",
    "rxns = get_reactions_of_met(me,'lipid_biomass', s=1, verbose=False)\n",
    "for rxn in rxns:\n",
    "    rxn.upper_bound = mu\n",
    "_=get_reactions_of_met(me,'lipid_biomass', s=1, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iter\tmuopt    \ta     \tb     \tmu1       \tstat1\n",
      "Finished compiling expressions in 212.524586 seconds\n",
      "Finished substituting S,lb,ub in 9.029733 seconds\n",
      "Finished makeME_LP in 1.079063 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "1 0.0 0.0 0.25 0.25 1\n",
      "Finished substituting S,lb,ub in 9.777054 seconds\n",
      "Finished makeME_LP in 1.075727 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "2 0.0 0.0 0.125 0.125 1\n",
      "Finished substituting S,lb,ub in 9.068559 seconds\n",
      "Finished makeME_LP in 1.088095 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "3 0.0625 0.0625 0.125 0.0625 optimal\n",
      "Finished substituting S,lb,ub in 8.932621 seconds\n",
      "Finished makeME_LP in 1.085652 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "4 0.09375 0.09375 0.125 0.09375 optimal\n",
      "Finished substituting S,lb,ub in 8.774667 seconds\n",
      "Finished makeME_LP in 1.093390 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "5 0.09375 0.09375 0.109375 0.109375 1\n",
      "Finished substituting S,lb,ub in 8.775150 seconds\n",
      "Finished makeME_LP in 1.082010 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "6 0.1015625 0.1015625 0.109375 0.1015625 optimal\n",
      "Finished substituting S,lb,ub in 8.823471 seconds\n",
      "Finished makeME_LP in 1.084315 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "7 0.1015625 0.1015625 0.10546875 0.10546875 1\n",
      "Finished substituting S,lb,ub in 8.915846 seconds\n",
      "Finished makeME_LP in 1.069650 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "8 0.1015625 0.1015625 0.103515625 0.103515625 1\n",
      "Finished substituting S,lb,ub in 8.841048 seconds\n",
      "Finished makeME_LP in 1.077709 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "9 0.1015625 0.1015625 0.1025390625 0.1025390625 1\n",
      "Finished substituting S,lb,ub in 8.690369 seconds\n",
      "Finished makeME_LP in 1.060216 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "10 0.1015625 0.1015625 0.10205078125 0.10205078125 1\n",
      "Finished substituting S,lb,ub in 8.878078 seconds\n",
      "Finished makeME_LP in 1.118633 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "11 0.1015625 0.1015625 0.101806640625 0.101806640625 1\n",
      "Finished substituting S,lb,ub in 8.948643 seconds\n",
      "Finished makeME_LP in 1.078310 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "12 0.1016845703125 0.1016845703125 0.101806640625 0.1016845703125 optimal\n",
      "Finished substituting S,lb,ub in 8.993294 seconds\n",
      "Finished makeME_LP in 1.082222 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "13 0.10174560546875 0.10174560546875 0.101806640625 0.10174560546875 optimal\n",
      "Finished substituting S,lb,ub in 9.033439 seconds\n",
      "Finished makeME_LP in 1.085481 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "14 0.101776123046875 0.101776123046875 0.101806640625 0.101776123046875 optimal\n",
      "Finished substituting S,lb,ub in 8.978117 seconds\n",
      "Finished makeME_LP in 1.222057 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "15 0.101776123046875 0.101776123046875 0.1017913818359375 0.1017913818359375 1\n",
      "Finished substituting S,lb,ub in 8.837475 seconds\n",
      "Finished makeME_LP in 1.080285 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "16 0.10178375244140625 0.10178375244140625 0.1017913818359375 0.10178375244140625 optimal\n",
      "Finished substituting S,lb,ub in 8.684090 seconds\n",
      "Finished makeME_LP in 1.071217 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "17 0.10178375244140625 0.10178375244140625 0.10178756713867188 0.10178756713867188 1\n",
      "Finished substituting S,lb,ub in 8.770837 seconds\n",
      "Finished makeME_LP in 1.067535 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "18 0.10178375244140625 0.10178375244140625 0.10178565979003906 0.10178565979003906 1\n",
      "Finished substituting S,lb,ub in 8.688938 seconds\n",
      "Finished makeME_LP in 1.064450 seconds\n",
      "Getting MINOS parameters from ME_NLP...\n",
      "19 0.10178470611572266 0.10178470611572266 0.10178565979003906 0.10178470611572266 optimal\n",
      "Bisection done in 574.385 seconds\n"
     ]
    }
   ],
   "source": [
    "solve_me_model(me, max_mu = 0.5, min_mu = .05, using_soplex=False, precision = 1e-6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./me_models/iYO844b.pickle\", \"wb\") as outfile:\n",
    "    pickle.dump(me, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Oxidized_c'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-9738f7df1daa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msave_json_me_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mme\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'./me_models/iBS.json'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/io/json.py\u001b[0m in \u001b[0;36msave_json_me_model\u001b[0;34m(model, file_name)\u001b[0m\n\u001b[1;32m     65\u001b[0m         \u001b[0mshould_close\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 67\u001b[0;31m     \u001b[0mmodel_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mme_model_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     68\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[0;31m# Confirm that dictionary representation of model adheres to JSONSCHEMA\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/io/dict.py\u001b[0m in \u001b[0;36mme_model_to_dict\u001b[0;34m(model)\u001b[0m\n\u001b[1;32m    258\u001b[0m         \u001b[0mreactions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_get_attribute_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreactions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'reaction'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    259\u001b[0m         process_data=_get_attribute_array(model.process_data,\n\u001b[0;32m--> 260\u001b[0;31m                                           'process_data'),\n\u001b[0m\u001b[1;32m    261\u001b[0m         \u001b[0mmetabolites\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_get_attribute_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetabolites\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'metabolite'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    262\u001b[0m         \u001b[0mglobal_info\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_get_global_info_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mglobal_info\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/io/dict.py\u001b[0m in \u001b[0;36m_get_attribute_array\u001b[0;34m(dictlist, type)\u001b[0m\n\u001b[1;32m    223\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0m_reaction_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreaction\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mreaction\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictlist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'process_data'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0m_process_data_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictlist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'metabolite'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0m_metabolite_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetabolite\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mmetabolite\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictlist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/io/dict.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    223\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0m_reaction_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreaction\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mreaction\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictlist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'process_data'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0m_process_data_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictlist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'metabolite'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0m_metabolite_to_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetabolite\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mmetabolite\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictlist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/io/dict.py\u001b[0m in \u001b[0;36m_process_data_to_dict\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    180\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mattribute\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_PROCESS_DATA_TYPE_DEPENDENCIES\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprocess_data_type\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    181\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mattribute\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mspecial_list\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 182\u001b[0;31m             \u001b[0mdata_attribute\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattribute\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    183\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m             \u001b[0mnew_process_data_type_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mattribute\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_fix_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata_attribute\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/core/processdata.py\u001b[0m in \u001b[0;36melement_contribution\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    216\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    217\u001b[0m             contribution = {i: v for i, v in\n\u001b[0;32m--> 218\u001b[0;31m                             \u001b[0miteritems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcalculate_element_contribution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    219\u001b[0m                             if v}\n\u001b[1;32m    220\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/UCSD/cobrame-master/cobrame/core/processdata.py\u001b[0m in \u001b[0;36mcalculate_element_contribution\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    251\u001b[0m         \u001b[0melements\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdefaultdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    252\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mmet\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcoefficient\u001b[0m \u001b[0;32min\u001b[0m \u001b[0miteritems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstoichiometry\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 253\u001b[0;31m             \u001b[0mmet_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetabolites\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    254\u001b[0m             \u001b[0;31m# elements lost in conversion are added to complex, protein, etc.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mmet_obj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0melements\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmet_obj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGenerictRNA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/cobrapy-0.5.11/cobra/core/DictList.py\u001b[0m in \u001b[0;36mget_by_id\u001b[0;34m(self, id)\u001b[0m\n\u001b[1;32m     47\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m         \u001b[0;34m\"\"\"return the element with a matching id\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 49\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mlist_attr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattribute\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Oxidized_c'"
     ]
    }
   ],
   "source": [
    "save_json_me_model(me, './me_models/iBS.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./me_models/final/iYO844b.pickle\", \"wb\") as outfile:\n",
    "    pickle.dump(me, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mets = ['cl', 'sheme', 'mg2', 'ni2', 'thmpp', 'cu', 'nad', 'ca2', 'fmn', 'zn2', 'pydx5p', 'pi', 'pheme',\n",
    "# 'bmocogdp', '4fe4s', 'mn2', 'cbl1', 'cu2', 'k', 'cobalt2', '2fe2s', 'fe2']\n",
    "mets = ['glc__D']\n",
    "met_ids = [met+'_c' for met in mets]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gap_mets = brute_force_check(me,met_ids,'biomass_dilution',0.02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "me_ = me.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from qminospy import me2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This adds exchange reactions for metabolites not contained in iJO1366\n",
    "# Some of these cannot be produced by the model so they are added here\n",
    "exchange_list = ['4fe4s_c','2fe2s_c','udcpp_c','pydx5p_c','3fe4s','cl_c','sheme_c','cu_c','mn2_c',\n",
    "                 'bmocogdp_c','dpm_c','thmpp_c','zn2_c','cbl1_c','cobalt2_c']\n",
    "\n",
    "for met_id in exchange_list:\n",
    "    met_base = met_id.split('_')[0]\n",
    "    \n",
    "    if hasattr(me.metabolites,'{}_c'.format(met_base)):\n",
    "        met_c = me.metabolites.get_by_id('{}_c'.format(met_base))\n",
    "    else:\n",
    "        met_c = cobrame.Metabolite('{}_c'.format(met_base))\n",
    "        \n",
    "    if hasattr(me.metabolites,'{}_e'.format(met_base)):\n",
    "        met_e = me.metabolites.get_by_id('{}_e'.format(met_base))\n",
    "    else:\n",
    "        met_e = cobrame.Metabolite('{}_e'.format(met_base))\n",
    "        \n",
    "    ex_id = 'EX_{}_e'.format(met_base)\n",
    "    if hasattr(me.reactions,ex_id):\n",
    "        ex_r = me.reactions.get_by_id(ex_id)\n",
    "    else:\n",
    "        ex_r = cobrame.MEReaction(ex_id)\n",
    "        ex_r.add_metabolites({met_e:-1})\n",
    "        me.add_reaction(ex_r)\n",
    "    t_r = cobrame.MEReaction('{}_t'.format(met_base))\n",
    "    t_r.add_metabolites({met_e:-1,met_c:1})\n",
    "    me.add_reaction(t_r)\n",
    "    ex_r.upper_bound = 0 # Avoid secretion of this metabolites. These reactions are\n",
    "                         # only meant to supply cofactors for which no biosynthetic \n",
    "                         # pathway has been defined in the model.\n",
    "    ex_r.lower_bound = -1000"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
